### DISCUSSION TITLE
Prior art for PSD?

### DISCUSSION THREAD

**Discussion by tantaman** [dexie/Dexie.js#1617] [General]
I noticed that Dexie patches Promises to enable async context and refers to this as "PSD" [here](https://github.com/dexie/Dexie.js/blob/d2c22ba983a1ab116e165993c47aa93ecd5646d6/src/helpers/promise.js#L13).

I haven't heard of context propagation referred to as PSD before nor can I find it with some googling. Are there some original resources of context prop and PSD that you can refer me to?

I'm interested in learning more as I have context propagation problems I need to solve myself.

**Comment by dfahlander** [comment]
Hi! When first implemented PSD in dexie, I was not aware of other context-preserving techniques, which is the reason for its proprietary naming which is unique to Dexie.js. The concept is more known as zones (like angular's zone.js) or contexts, as known in libraries such as react, vue, svelte and solidjs. Dexie's PSD context is unique compared to other libraries in that it is preserved across async calls.

**Reply by tantaman** [reply]
Are there some docs on how to correctly use `newScope` to attach a context to an async function?

```
const result = await newScope(someAsyncFunc, myContextProps);
```

I've used it with some success (figuring out to call incrementExpectedAwaits and decrementExpectedAwaits) but have had some issues where the scope is sometimes still active even after awaiting the async task.

I've been deriving my usage from examples like this one from the dexie codebase:
https://github.com/dexie/Dexie.js/blob/master/src/classes/version/schema-helpers.ts#L150-L169

**Reply by tantaman** [reply]
I seem to have gotten it working correctly. Will follow up with a link to the various unit tests.

I split the `zone` code out of Dexie for use in other projects if you're interested. My current use case is for [software transactional memory (STM) in JS](https://github.com/aphrodite-sh/vulcan/tree/main/ts/packages/value).
